### ISSUE TITLE
Generator for tests

### ISSUE THREAD

**Issue by guygastineau** [exercism/bash#185]
Hi all,
## I'm currently developing a Ruby and ERB based generator to make the test files for our problems based on a template and the data from problem specifications.  I would say I'm in middle alpha stage.
1. Successfully generates tests when everything is in the same directory
2. Successfully generates tests when the template is in another file.
#### Currently I'm ironing out a couple things to make it friendlier to contributors who will use this
- In the end I plan on using a shell wrapper for the ruby program
- I will need to generate in depth documentation so that implementation of generator templates for exercises can be done by any contributor.
- to simplify the usage and creation of such tools I think we should create the following **conventions**:
  - scripts that involve this repo and other repos extended from exercism should expect all the repos to be in parallel directories.  This is expected in the ruby track and specified in their [README.md](https://github.com/guygastineau/ruby/blob/master/README.md) in the *regenerating a test suite* section.
    - This would allow scripts in the bash track to be shorter and more straight forward.
  - Documentation of developement tools should be comprehensive - especially if the tool is not written exclusively in bash.
    - This generator project will be mostly ruby and I think it is appropriate for documentation on it to make it as accessible as possible even to those unfamiliar with ruby.

**Comment by budmc29** [comment]
I'll ask the first sensible question to ask: Do we need this?

**Comment by guygastineau** [comment]
@budmc29 I think that is great question.
#### A couple reasons why I think we probably do need it:
1. Running [`canonical_data_check.sh`](https://github.com/guygastineau/bash/blob/master/scripts/canonical_data_check.sh) shows about half or more of our exercises don't have a version file in their `.meta/`.  Many of these are probably out of sync with the problem-specifications repo.
2. Currently the only way we have to bring an excercise into compliance with its `canonical-data.json` is by manually updating the test.  This is cumbersome and time consuming at best, and given that we don't have a lot of contributors in this repo full compliance will likely never happen.
3. I think we all want this track to grow in terms of both track members and exercises.  I think more exercises will also bring more members, but as the exercises grow so will the time cost of keeping everything up to date.
4. I think @sjwarner-bp has done a great a thing by making the `canonical_data_check.sh` to show us what needs updating.  However, the output from that script currently amounts to a list of things that are *probably never going to happen*.
  - I want to be clear that the above is not a jab at the maintaners. You all have done incredible work, and I can see most of you have contributed over 1,000 lines of code.  I appreciate your ongoing efforts!
  - I am simply trying to be realistic about the amount of work we can collectively accomplish manually.
  - Since exercises are basically the bricks that build our track, I think that a semi-automated solution exercise maintenance is practically a necessity.
5. We have no guarantee that `problem/specification` compliance or a test generating framework wont be enforced for all tracks in future.
  - Getting this solution set up and functional now means getting ahead of the curve if/when enforcement ever becomes a reality.
6. It is generally cool to have tools we have built ourselves to help maintain the track.  We are after all using exercism.io to become better programmers and engineers.

**Comment by guygastineau** [comment]
UPDATE: 
- I'm having a conversation with @kotp on slack about how much of this can and should be kept in bash only.
- I think a good purpose for this issue now would be to continue discussing the question posed by @budmc29.
  - Essentially, should we implement a generator? (I've given my reasons in support of implementation above.)
  - There's no need to worry now about whether or not it is my ruby project.
-------------------------------------------------------------------------------------------------------------------------------------------
#### If we decide to implement a test generator
- We need to decide how much must be bash
- What portion (if any) is written in another language, and what that language would be.
- We need to decide how it will integrate with tools (existing and yet to come) for managing our exercise versions.
- Anything else we can think of.

**Comment by budmc29** [comment]
Thanks for taking the time to propose this @guygastineau, looks well thought out and I understand how this will be able to help us in the future.

From what I can see you are truly excited about this tool, and if you have enough time to work on this, and we can have real applications to our track, I see no issue with adding the tool to the track.

The only requirement for it, that I truly believe should not be overlooked, is that any tool we use should be written in bash.
Personally I wouldn't feel good to be a contributor in the Ruby track and have someone write a tool in Scala, no matter if that's the best language for said tool, and that it has great comments.

**Comment by guygastineau** [comment]
Thanks for the response @budmc29.

#### Doing this with a just bash.
- @kotp informed me that the Delphi track did something similarly whereby they took inspiration from the ruby track and implemented a generator framework built entirely in Delphi.
- I assume `jq` commands will be necessary to get the information out of the `canonical-data.json` file.  I am unaware of another way to parse json with pure bash - exept maybe some really complicated processing with the read command: I'll look into this too.
- In bash I assume there will be a a parent script that runs with an exercise slug as an argument.  The generator template in the exercise directory will probably need to be a function.
  - in that way, our 'templates' will be a bit more complicated and harder to write, but it should be possible to source them in as functions and apply them with the parent script.

- I would suggest that, following the style guide, all of our executable scripts not have the `.sh` extension, but libraries - including these generator templates - would need the `.sh` extension and should be left read obly whereas the executables will be read and execute only. Obviously, if someone is changing them locally they will need to use chmod.

----------------------------------------------------------------------
#### What do you all think about this precursory outline?

- Honestly I started the project with Ruby because I started leaning ERB and it was just so easy to use, but I like the idea of implementi g this in bash for the sort of weird coolness factor it will have.

**Comment by budmc29** [comment]
I don't want to cut on your enthusiasm, that's a great thing to have, so let's see what we can do.

You can go two ways about this, write the tool in Ruby, make it work well, and then create PR's with the results and that will help the track. If the tool is good and helps, I will start using it as well since I am a Ruby developer, and I assume others might want to use it, but I don't want to add it to the track, and the the CI pipeline since it's not the correct tool.

This has a benefit that you can propose using the tool on the Ruby track, as it is, thought from my observations in most of the tracks we don't lack tools to do any work, we just like the time to do the actual work.

If you still consider this a good idea after that, you can rewrite it in Bash, having all the experience and lessons learned from building it with Ruby, and having figured out all the unknowns.

Or you can start writing the tool in Bash, and do the same process as before, and if the tool is easy to use, integrates well and it's not dependent on one developer to maintain it, I'll be more than happy to add it to the CI pipeline and I assume other maintainers might feel the same.

**Comment by guygastineau** [comment]
@budmc29 yeah, since it is functional in Ruby right now I thought I might go ahead and PR updated tests in the coming weeks while I work on the bash version.

I'll keep you updated on my progress, and when the bash version starts to get somewhere I will probably put it on a branch in my fork so you all can see/contribute to it's developement.

**Comment by guygastineau** [comment]
After working on this in bash today I believe that the template functions will be far to involved.
I have been doing it with jq.  Basically, each template implemented will require quite an advanced implementation of jq in order to get everything out of the .json and related correctly for insertion into the test file.

#### It is my opinion that not involving an OOP language that has libraries for json allowing conversion of json into structs within that language is not only very cumbersome but will in fact be much harder than a contributor learning how to write a template that works with another language.

**Comment by guygastineau** [comment]
#### So I thought you all might be interested in seeing my progress in refactoring the ruby version of a generator.

- I pushed the branch to my fork [here.](https://github.com/guygastineau/bash/tree/test-generator)
- It includes:
  - `bin/generator_test` a wrapper *unfinished but functional*
  - `lib/generator_class.rb` *new directory `lib` for the files that are technically libs*
  - `scripts/test_generator.rb` *the ruby script that actually invokes the generator_class objects*
  - `exercises/atbash-cipher/.meta/generator_template.erb` *so I could test the functionality of this tool*
  - The resulting changes in `atbash_cipher_test.sh` and the addition of `old_test.sh` are the results of running the program with the wrapper `bin/generate_test`.
- I would intend for users to delete the old test once it is verified that the new test is formatted fine and that the example.sh passes all the tests.

-------------------------------------------------------------------------------------------------------------------------------------------

- Ultimately I think the wrapper should check that the version file is even with the problem-specifications version, but I think it should be changed manually to avoid any issues.
- The wrapper also runs the `bin/validate-exercises` script at the end and gives instructions on how to tidy up and trouble shoot.  This part should be in a function but it isn't for now.  It shoould also be expanded.
- Ultimately I think the wrapper should also catch errors from the ruby program, provide useful error messages about any errors caught, and revert the state of things with `mv old_test.sh <original_name_of_test>`.

**Comment by guygastineau** [comment]
@budmc29

#### I think I see what you meant in your last comment that others might want to use it.
- I'm understanding:
  - contributors already familiar with Ruby might like to use it.
    - If it's good and simple.
  - It would be inappropriate to add it to the CI *pipeline?*(I'm a total boob at Travis).
    -I agree, it shouldn't be a required component for developement in the track.

- I'm confused about:
  - Should I should PR the fork I linked?
    - Would it still go in the main repo, and just not be integrated with certain core systems?
  - Is there a preferred way for me to point other contributors to this code if I can assume they are interested?

----------------------------------------------------------------------
#### I would like to:
- Get a workable solution for test generation available to any contributors to this track who want to use it .
- Version and refine the project if  possible after some use (*Refactoring for improvements?*).
- Get the tool to the point where those who use it enjoy it and agree with it's structure and conventions.
  - I think more templates will make it more fun and useful.
  - I am willing to write a lot of templates.
    - I will probably make templates for the out of date ones as returned by `bin/canonical_data_check.sh` first.
    - That way I can help get `meta/version` files for all of our exercises.

**Comment by budmc29** [comment]
You can create a PR if you want, but my opinion is that you should first prove the worth of the tool to the track and other maintainers, before trying to integrate it.

I would argue that you should keep the tool on a new repository on your account and point whoever is interested to that, rather than add it on this repo.

After all this talking about this, I'd be happy to see some actual results, for example a PR to update the existing exercises to be in line with the canonical-data.

**Comment by guygastineau** [comment]
Hey @budmc29,

Thanks for the info. Yeah, I figured I shouldn't just PR it to the repo. Thanks for giving me some solutions for showing it you all.

- I will put it isolated in a new repo.

- I will build templates for some out of date tests and PR the results.

- @kotp noticed that I had some trailing whitespace that needs to be gone.

- Also per his suggestion, I will be verifying the earliest version of ruby with which this works.

**Comment by budmc29** [comment]
Offtrack

I'm using this for showing me whitespace in vim
```
" highlight whitespace
highlight ExtraWhitespace ctermbg=red guibg=red
match ExtraWhitespace /\s\+$/
autocmd BufWinEnter * match ExtraWhitespace /\s\+$/
autocmd InsertEnter * match ExtraWhitespace /\s\+\%#\@<!$/
autocmd InsertLeave * match ExtraWhitespace /\s\+$/
autocmd BufWinLeave * call clearmatches()
```

and this will automatically remove whitespace in certain files
```
" remove trailing whitespace on save
fun! <SID>StripTrailingWhitespaces()
let l = line(".")
    let c = col(".")
    %s/\s\+$//e
call cursor(l, c)
endfun

autocmd FileType c,cpp,java,php,ruby,python autocmd BufWritePre <buffer> :call <SID>StripTrailingWhitespaces()
```

I have tried plugins but found this to be more reliable.

**Comment by guygastineau** [comment]
I put it up [here](https://github.com/guygastineau/bash-track-test-generator) by itself.
- I will find some solutions that are out of date and PR the results later today or tomorrow.

-------------------------------------------------------------------------------------------------------------------------------------------
#### Cool vim solution.
- I just installed installed [this](https://github.com/bronson/vim-trailing-whitespace) plugin to do it.

**Comment by guygastineau** [comment]
I PR'd an update to luhn made with my generator #186.

- I forgot to add the skips in the template, so I added them by hand.
- It is trivially easy to add the skips to the template and then comment the first one out though.

**Comment by guygastineau** [comment]
I will wait for each test update PR to go through  before PRing new updates, but I will go ahaead and make more templates and update tests adding them to the repository that holds the generator.

**Comment by guygastineau** [comment]
## Would people be more in favor of this tool being hosted in a separate repo.
- I can make it fetch the most recent canonical-data.json from the bash master repo.
- Then the tests could be generated in the directory for the program and PR'd seperately.
- I might make it fetch the example.sh too so that the program can check to see if the example.sh needs updates to conform to the new test.

#### Let me know what you all think!

**Comment by kotp** [comment]
Host it here, I would say.  It makes it visible.  If it is later ported, in whatever fashion, it is here.  It address the track, and is a tool that may not be mandatory, but is useful, so if accepted, it belongs here.

At least that is my vote.

**Comment by budmc29** [comment]
We can add the link to the tool in a readme file to make it visible to the users.

I think it will be cleaner to have something that's optional and not in the same language as our track in a separate repository.
This will also help with having good documentation and tests for the tool without adding too much noise to this track.

**Comment by guygastineau** [comment]
@kotp @budmc29

#### Thank you both for your input.
- For now I will work on the program in it's separate repo [here](https://github.com/guygastineau/bash-track-test-generator) 
  - As of now I'm using it on a branch of `exercism/bash`.
  - I am reworking it so it can work from that repo by pulling the json from the internet with 'open-uri', and I will be changing the path dependencies in favor of a model that works in this separate repo.
- I think this way people can use this tool and build up templates and docs without cluttering the repo.
- When everything is more established it there will be more important discussions to have.
  - Notably, the process of making PR's to update tests is showing me that our tests are not standardized.
  - I think the coming week or so of test updating PR's from me will force much standardization into our tests, as it already has to some degree.
- I think it will also be prudent after some further development to get the input of some other maintainers for this track.
  - It would be nice to have at least half of the maintainers opinions for making a decision about where to house this project and how to document it.

#### I will post here again:
- When I've added fetching the json from the internet.
  - Hopefully it will be enticing for others to use once it no longer needs to be put in a dead end branch of the main repo in order to generate tests.

**Comment by guygastineau** [comment]
Reworked the project.  It is actually just ruby now.
It functions without the presence of the bash track repo now.
Feel free to take a look [here](https://github.com/guygastineau/bash-track-test-generator).

**Comment by budmc29** [comment]
Sounds great.

**Comment by guygastineau** [comment]
Hey @budmc29 what do you think about chanigng our test example and the template [here](https://github.com/exercism/bash/blob/master/_template/exercise_slug_test.sh) so that they don't have the newline inbetween the run command and the expected status and ouput?

I think most of our exercises now don't have the newline and the main repo for BATS, [here](https://github.com/sstephenson/bats), also doesn't use the newline.

I'm thinking this means it is more idiomatic for BATS not to have a newline there.
